### PR TITLE
Feature: interpolate: Extrapolate left/right

### LINKED ISSUE #1255: interpolate: Extrapolate left/right
## Description

I'd like to see extrapolating for both sides, just like in v1.

## Motivation

Currently we can only extrapolate left and right at the same time, but if I wanted to clamp left and extend identity for the right value, that wouldn't be possible. I'd have to use a workaround: Use no extrapolation and then use `Math.min` or `Math.max` for "clamping".

## Code example

A solution for this would be:

```ts
const x = interpolate(scrollX, [0, HEADER_HEIGHT, contentSize.x], [0, 0, contentSize.x + HEADER_HEIGHT], { extrapolateLeft: 'clamp', extrapolateRight: 'identity' })
```

To make it backwards compatible (and also cleaner if you don't need different extrapolations) the `interpolate` function could take both types:

```ts
type ExtrapolateParameter = Extrapolate | { extrapolateLeft?: Extrapolate, extrapolateRight?: Extrapolate } | undefined 
```

### PR DESCRIPTION
**Pull request by Roka20012**
## Wix Goodness Squad Event

Motivation: to see extrapolating for both sides, just like in v1.

Fixes #1255

## Changes

- Added ts types for extrapolation
- Added ExtrapolationExample
- Added Error handling for invalid config or invalid interpolation value 

## Live example

IDK which example is better currently 2 in this PR

https://user-images.githubusercontent.com/32239452/102691648-a03d0780-4216-11eb-90e7-7ef8a5f1590a.mov

---
https://user-images.githubusercontent.com/32239452/102694608-df755380-422a-11eb-839c-8872d1427807.mov

## Checklist

- [x] Included code example that can be used to test this change
- [x] Updated TS types

### CODE REVIEW

**Comment by Roka20012** [comment]
@terrysahaidak sup dude!
I think this PR need to review

Thanks again for being a maintainer on the Wix Goodness Squad Event

Cheers ✌️

**Comment by terrysahaidak** [comment]
Hey, @karol-bisztyga this is yet another PR as a result of our event, could you please assign someone for review. Thanks.

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:29]: Could you please invert the logic here?

```
value === Extrapolate.EXTEND ||
value === Extrapolate.CLAMP ||
value === Extrapolate.IDENTITY
```

is way easier to read.

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:55]: Won't `isExtrapolate` return `false` for `null` / `undefined`?

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:99]: Why it's commented out? If we need default then it should be left and if we don't please just remove this value.

BTW, does it break backward compat?

**Review by jakub-gonet — COMMENTED** [review]
Left some comments, but somebody from turboteam should review this.

File: src/reanimated2/interpolation.js
Comment (karol-bisztyga) [inline src/reanimated2/interpolation.js:78]: As for the typeof checks: checking for type `object` occurs a several times. I'd move those checks to the end of this function and then bring it before them so it would look like this:
```
f() {
	// ... some checks
	if (typeof type !== 'object') {
		return;
	}
	// ... all the checks that contained `typeof type !== 'object'` before
}
```

File: src/reanimated2/interpolation.js
Comment (karol-bisztyga) [inline src/reanimated2/interpolation.js:98]:
```suggestion
  // e.g. function interpolate(x, input, output, type = Extrapolate.CLAMP)
```

Comment (karol-bisztyga) [reply src/reanimated2/interpolation.js:98]: please fix this typo here 🙏

**Comment by karol-bisztyga** [comment]
It's a solid PR, I'll be happy to merge it, just please make those couple of thing more clear and I think we're good to go 😉

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Yeap it will

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: > Why it's commented out? If we need default then it should be left and if we don't please just remove this value.

Removed

---

> BTW, does it break backward compat?

If you mean using it like this -> ```interpolate(translation.y.value, [0, -75], [0, -150], 'clamp')``` 

It will work like before 🙂

File: Example/src/App.js
Comment (Roka20012) [reply Example/src/App.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

**Comment by Roka20012** [comment]
@karol-bisztyga @jakub-gonet please re-check I have made some changes for this PR ✌️ ☃️

File: src/reanimated2/interpolation.js
Comment (karol-bisztyga) [inline src/reanimated2/interpolation.js:41]:
```suggestion
      }) or interpolate(value, [inputRange], [outputRange], 'clamp')`
```

File: src/reanimated2/interpolation.js
Comment (karol-bisztyga) [inline src/reanimated2/interpolation.js:51]: Maybe it could be done using less code?
How about:
```
type?.constructor === Object &&
    ((Object.keys(type).length >= 1 &&
      !('extrapolateLeft' in type && 'extrapolateRight' in type)) ||
    Object.keys(type).length > 2)
```
or
```
type?.constructor === Object &&
    (([1,2].includes(Object.keys(type).length) &&
      !('extrapolateLeft' in type && 'extrapolateRight' in type)) ||
    Object.keys(type).length > 2)
```

File: src/reanimated2/interpolation.js
Comment (karol-bisztyga) [inline src/reanimated2/interpolation.js:106]: I don't think that ternary replaces the original conditions, can you revert this? Or am I missing something?

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js:51]: Yep, looks too much anyway I ended up with this

Your code will not work for this code

In case if one of `extrapolateLeft | extrapolateRight` field is missing (by default if one of them is missing it's `extrapolate === Extrapolate.EXTEND`)

```js
interpolate(translation.y.value, [0, -75], [0, -150], {
        extrapolateLeft: 'clamp',
})
```

Comment (karol-bisztyga) [reply src/reanimated2/interpolation.js:51]: Yes, you are right, sorry, I missed one thing in those conditions, it's ok.

**Review by karol-bisztyga — APPROVED** [review]

**Comment by karol-bisztyga** [comment]
For me, it looks OK. @jakub-gonet?

**Review by piaskowyk — APPROVED** [review]

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:35]: Isn't that just `type === null`?

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:54]:
```suggestion
      `Reanimated: config object is not valid, please provide valid config, for example:
```

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:103]: Just making sure, it can be null here? (`typeof null === "object"`)

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:46]: 1. Safe navigation operator isn't necessary, we checked for null earlier.
    From [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor#description):
    > All objects (with the exception of objects created with Object.create(null)) will have a constructor property. 
2. Why we check for constructor property here?

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:48]: Nit; `in` operator traverses prototype hierarchy, we should probably use [`type.hasOwnProperty('extrapolateLeft')`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty).

Just a nit, because we expect an object as a dictionary here.

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:76]: I'd like to have this and other error messages as string constants:
```js
const EXTRAPOLATE_RIGHT_ERROR_MSG = `(...)`;
```

File: react-native-reanimated.d.ts
Comment (jakub-gonet) [inline react-native-reanimated.d.ts:55]: You already handle undefined in `extrapolate?`

**Review by jakub-gonet — COMMENTED** [review]
Left a few comments.

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js:35]: It's for `false` and empty string as well or false value

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js:103]: That's why I am doing `validateType` on `101` line

Comment (jakub-gonet) [reply src/reanimated2/interpolation.js:103]: Makes sense, thanks for explaining.

Line 101 if line numeration changes:
`validateType(type);`

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [reply src/reanimated2/interpolation.js:35]: Oh, right.

File: react-native-reanimated.d.ts
Comment (Roka20012) [reply react-native-reanimated.d.ts:55]: If you mean `interpolate(translation.y.value, [0, -75], [0, -150], undefined})`

It will work like `Extrapolate.EXTEND`

If you mean `interpolate(translation.y.value, [0, -75], [0, -150], {
        extrapolateLeft: undefined,
})`

It should throw an error

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Good catch, removed ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js]: Done ✅

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js:48]: Oh actually just a `!type` would be enough)

File: src/reanimated2/interpolation.js
Comment (jakub-gonet) [inline src/reanimated2/interpolation.js:61]: Why not do it like here ⬇️?
```suggestion
  const hasExtrapolateLeft = type.hasOwnProperty('extrapolateLeft');
  
  const hasExtrapolateRight = type.hasOwnProperty('extrapolateRight');
```

**Review by jakub-gonet — APPROVED** [review]
LGTM, left one comment.

File: react-native-reanimated.d.ts
Comment (jakub-gonet) [reply react-native-reanimated.d.ts:55]: I meant that we have this interface:
```ts
    interface InterpolationConfig {
      inputRange: ReadonlyArray<Adaptable<number>>;
      outputRange: ReadonlyArray<Adaptable<number | string>>;
      extrapolate?: ExtrapolateParameter; // <- extrapolate is optional
      extrapolateLeft?: Extrapolate;
      extrapolateRight?: Extrapolate;
    }
```

and this type:
```ts
type ExtrapolateParameter = Extrapolate | { extrapolateLeft?: Extrapolate, extrapolateRight?: Extrapolate } | undefined 
```

Marking extrapolate as optional already adds `| undefined` to the value.

File: src/reanimated2/interpolation.js
Comment (Roka20012) [reply src/reanimated2/interpolation.js:61]: Because of eslint error 😕

Comment (jakub-gonet) [reply src/reanimated2/interpolation.js]: I assume this is caused by this rule:
https://eslint.org/docs/rules/no-prototype-builtins

We don't use `Object.create` nor we shadow prototype methods, so it was safe to ignore.

File: react-native-reanimated.d.ts
Comment (Roka20012) [reply react-native-reanimated.d.ts]: Nice catch, fixed ✅

Removed  `| undefined ` from `ExtrapolateParameter `

**Comment by Roka20012** [comment]
@karol-bisztyga thanks for merging!❤️

**Pull request merged**
